CopilotKit/packages/runtime/src/graphql/resolvers/copilot.resolver.ts

```diff
@@ -48,15 +135,43 @@ export class CopilotResolver {
     const response = {
       threadId,
       runId,
-      status: firstValueFrom(responseStatus),
+      status: firstValueFrom(responseStatus$),
       messages: new Repeater(async (pushMessage, stopStreamingMessages) => {
+        if (data.cloud?.guardrails) {
+          invokeGuardrails({
+            baseUrl: copilotCloudBaseUrl,
+            copilotCloudPublicApiKey,
+            data,
+            onResult: (result) => {
+              guardrailsResult$.next(result);
+              if (result.status === "denied") {
+                responseStatus$.next(
+                  plainToInstance(FailedResponseStatus, { reason: result.reason }),
+                );
+                interruptStreaming$.next();
+              }
+            },
+          });
+        }
+
+        let eventStreamSubscription: Subscription;
+
         // run and process the event stream
-        const eventStream = eventSource.process(copilotRuntime.actions).pipe(
-          // shareReplay() ensures that later subscribers will see the whole stream instead of
-          // just the events that were emitted after the subscriber was added.
-          shareReplay(),
-        );
-        eventStream.subscribe({
+        const eventStream = eventSource
+          .process({
+            serversideActions: copilotRuntime.actions,
+            guardrailsResult$: data.cloud?.guardrails ? guardrailsResult$ : null,
+          })
+          .pipe(
+            // shareReplay() ensures that later subscribers will see the whole stream instead of
+            // just the events that were emitted after the subscriber was added.
+            shareReplay(),
+            finalize(() => {
```

Comment:
Is `finalize()` called once 100%?

Reply:
Not sure, but even if it's called multiple times, it's not an issue. This was the missing piece to solve the hanging streams for when `.complete()` isn't called.